File: 1252.cells-with-odd-values-in-a-matrix.cpp
/*
 * @lc app=leetcode id=1252 lang=cpp
 *
 * [1252] Cells with Odd Values in a Matrix
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int oddCells(int m, int n, vector<vector<int>>& indices) {
    bool is_odd_row[m];
    bool is_odd_col[n];
    memset(is_odd_row, 0, sizeof(is_odd_row));
    memset(is_odd_col, 0, sizeof(is_odd_col));
    for (const auto& index : indices) {
      is_odd_row[index[0]] = !is_odd_row[index[0]];
      is_odd_col[index[1]] = !is_odd_col[index[1]];
    }

    int odd_row = 0;
    for (int i = 0; i < m; i++) {
      if (is_odd_row[i]) {
        odd_row++;
      }
    }
    int odd_col = 0;
    for (int i = 0; i < n; i++) {
      if (is_odd_col[i]) {
        odd_col++;
      }
    }
    return odd_row * n + odd_col * m - 2 * odd_row * odd_col;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<int>> indices;
  indices = {{1, 1}, {0, 0}};
  ASSERT_EQ(s.oddCells(2, 2, indices), 0);
  indices = {{0, 1}, {1, 1}};
  ASSERT_EQ(s.oddCells(2, 3, indices), 6);
  return 0;
}

File: 1252.cells-with-odd-values-in-a-matrix.rs
/*
 * @lc app=leetcode id=1252 lang=rust
 *
 * [1252] Cells with Odd Values in a Matrix
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn odd_cells(m: i32, n: i32, indices: Vec<Vec<i32>>) -> i32 {
        let mut rows = vec![0; m as usize];
        let mut cols = vec![0; n as usize];
        for idx in &indices {
            rows[idx[0] as usize] += 1;
            cols[idx[1] as usize] += 1;
        }
        let mut ret = 0;
        for row in 0..m {
            for col in 0..n {
                ret += (rows[row as usize] + cols[col as usize]) % 2;
            }
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::odd_cells(2, 3, vec![vec![0, 1], vec![1, 1]]), 6);
}

#[test]
fn test2() {
    assert_eq!(Solution::odd_cells(2, 2, vec![vec![1, 1], vec![0, 0]]), 0);
}
